Fixes #29654: take reference to RWLayer while committing/exporting

yummypeng opened the pull request:
fixes #29654

Take an extra reference to rwlayer while the container is being
committed or exported to avoid the removal of that layer.

Also add some checks before commit/export.

@coolljt0725 @thaJeztah @cpuguy83 PTAL :stuck_out_tongue_winking_eye: 

Signed-off-by: Yuanhong Peng <pengyuanhong@huawei.com>

aaronlehmann commented on layer/layer.go:
```diff
@@ -188,6 +188,8 @@ type Store interface {
 	GetMountID(id string) (string, error)
 	ReleaseRWLayer(RWLayer) ([]Metadata, error)
+	LockRWLayer()
+	UnlockRWLayer()
```
ping @dmcgowan 

I'm not sure this interface change is a good idea. It feels like it breaks the abstraction.

It's been awhile since I looked at this code, but would it make sense for `RWLayer`'s `TarStream` method to lock the layer while the stream is in progress?

dmcgowan replied:
I agree, this is not a suitable addition to the interface.

If `TarStream` needs to block other operations it should hold a lock that blocks other operations until it is completed the stream.

yummypeng replied:
@aaronlehmann I've tried to lock the rw layer just in the progress of `TarStream` but found it was not enough. In other words, #29654 was still not resolved.

I found that the dm device would be busy as long as it is mounted. That's why I added these two interfaces (I knew it was a bit ugly though).

cpuguy83 replied:
@yummypeng Wouldn't you need to lock during for each invocation of `ls.driver.Remove`?

yummypeng replied:
@cpuguy83  `daemon.layerStore.ReleaseRWLayer(container.RWLayer)` has already been locked with the same lock.

dmcgowan commented:
@yummypeng relying on holding `layerStore.mountL` is not going to be an acceptable solution as it would block any other `RWLayer` creation while it is being called. As @cpuguy83 suggested, we a lock that effects `Remove`, to block removing of an active layer. Maybe we add a rw mutex to `mountLayer` and hold it during certain operations and block unmounting/removal of that layer until that lock can be acquired for write.

cpuguy83 commented:
Any update here?

yummypeng commented:
@cpuguy83 Sorry, I'm busy with other things recently. I will update this pr soon.

cpuguy83 commented:
Thanks, no worries!

yummypeng commented:
Sooorry for late update. :angel:
@cpuguy83 @dmcgowan  Since it's difficult get the rw mutex in `mountLayer` in package `daemon`, so I have to add two methods (`RLockRWLayer()` & `RUnlockRWLayer()`) to `RWLayer` which seems not so suitable. Any suggestions?

cpuguy83 commented:
@yummypeng This is locking at too high a level.
The daemon should not be concerned about layer locking.

I haven't looked at this much, but calls to a layer's `TarStream()` function should take a reference and then closing the produced `io.ReadCloser` should remove that reference.
This would prevent unmounting, removing, etc.

cpuguy83 commented:
* take a reference, or lock a mutex... something to look into.

yummypeng commented:
> This is locking at too high a level.

@cpuguy83 As long as the rwLayer of a container is mounted, the removal of that layer would be forbidden. So I think we should lock the rwLayer of a container before `daemon.Mount(container)` and unlock it after `daemon.Unmount(container)`. I can move the lock into `exportContainerRw` and `containerExport` if it's acceptable.

cpuguy83 commented:
@yummypeng The handling of mounting/unmounting is done in the layer store. This should not touch daemon at all, it's not the daemon's concern.

yummypeng commented:
> As long as the rwLayer of a container is mounted, the removal of that layer would be forbidden. So I think we should lock the rwLayer of a container before daemon.Mount(container) and unlock it after daemon.Unmount(container).

@cpuguy83 I've got what you meant, but if the above is true, it's expected that we lock a mutex before #L274 and unlock it after #L285. (take the example of commit)

```go
273 func (daemon *Daemon) exportContainerRw(container *container.Container) (io.ReadCloser, error) {
274         if err := daemon.Mount(container); err != nil {
275                 return nil, err
276         }
277
278         archive, err := container.RWLayer.TarStream()
279         if err != nil {
280                 daemon.Unmount(container) // logging is already handled in the `Unmount` function
281                 return nil, err
282         }
283         return ioutils.NewReadCloserWrapper(archive, func() error {
284                         archive.Close()
285                         return container.RWLayer.Unmount()
286                 }),
287                 nil
288 }
```

tonistiigi commented:
It seems to me that this could be fixed by just having an extra reference to rwlayer. Like in https://github.com/docker/docker/pull/31159/files#diff-57537ed135debc1c76b8f5543330452aR16

yummypeng commented:
@tonistiigi Thanks for suggestion. :heart:  But I think taking a reference is not the solution as @cpuguy83 
 suggested in [#29743 (comment)](https://github.com/docker/docker/pull/29743#issuecomment-269631388) because reference will never wait for something to complete.

tonistiigi commented:
@yummypeng Is it required that it should wait? Isn't the expected behavior just that it would not cause committing/exporting to error.

cpuguy83 commented:
I don't think it's required to wait unless, but then we'd need to error with "sorry this is busy right now"

cpuguy83 commented:
Or automatically cleanup with it's not busy.

tonistiigi commented:
@cpuguy83 If you take the reference it would just clean up after commit/export finishes. Same as in #32293

yummypeng commented:
@tonistiigi @cpuguy83 Updated :stuck_out_tongue_winking_eye:

cpuguy83 commented on layer/layer_store.go:
```diff
@@ -563,7 +563,7 @@ func (ls *layerStore) ReleaseRWLayer(l RWLayer) ([]Metadata, error) {
 
 	if m.hasReferences() {
-		return []Metadata{}, nil
+		return []Metadata{}, ErrActiveMount
```
I'm not sure that we really need to return an error here.
At this point the layer is released, just not cleaned up. I think that's ok.

yummypeng replied:
Yes, I agree with you. But when we delete a container, if the rwlayer is still busy, we should not do the `defer func()` in #L116, that is to say, we need an error return from `daemon.layerStore.ReleaseRWLayer`.
```go
 80 func (daemon *Daemon) cleanupContainer(container *container.Container, forceRemove, removeVolume bool) (err error) {
...
114         // If force removal is required, delete container from various
115         // indexes even if removal failed.
116         defer func() {
117                 if err == nil || forceRemove {
118                         daemon.nameIndex.Delete(container.ID)
119                         daemon.linkIndex.delete(container)
120                         selinuxFreeLxcContexts(container.ProcessLabel)
121                         daemon.idIndex.Delete(container.ID)
122                         daemon.containers.Delete(container.ID)
123                         if e := daemon.removeMountPoints(container, removeVolume); e != nil {
124                                 logrus.Error(e)
125                         }
126                         daemon.LogContainerEvent(container, "destroy")
127                 }
128         }()
129
130         if err = os.RemoveAll(container.Root); err != nil {
131                 return fmt.Errorf("Unable to remove filesystem for %v: %v", container.ID, err)
132         }
133
134         // When container creation fails and `RWLayer` has not been created yet, we
135         // do not call `ReleaseRWLayer`
136         if container.RWLayer != nil {
137                 metadata, err := daemon.layerStore.ReleaseRWLayer(container.RWLayer)
138                 layer.LogReleaseMetadata(metadata)
139                 if err != nil && err != layer.ErrMountDoesNotExist {
140                         if err == layer.ErrActiveMount {
141                                 return fmt.Errorf("Sorry, the rwlayer of container %v is busy right now", container.ID)
142                         }
143                         return fmt.Errorf("Driver %s failed to remove root filesystem %s: %s", daemon.GraphDriverName(), container.ID, err)
144                 }
145         }
146
147         return nil
148 }
```

thaJeztah replied:
ping @cpuguy83 ^^

cpuguy83 replied:
This defer is gone now.

Commit:
Fixes #29654: take reference to RWLayer while committing/exporting

Take an extra reference to rwlayer while the container is being
committed or exported to avoid the removal of that layer.

Also add some checks before commit/export.

Signed-off-by: Yuanhong Peng <pengyuanhong@huawei.com>

yummypeng commented:
rebased :wink:  @cpuguy83 @thaJeztah PTAL

cpuguy83 commented on daemon/export.go:
```diff
@@ -36,6 +47,11 @@ func (daemon *Daemon) ContainerExport(name string, out io.Writer) error {
 }
 
 func (daemon *Daemon) containerExport(container *container.Container) (io.ReadCloser, error) {
+	rwlayer, err := daemon.layerStore.GetRWLayer(container.ID)
```
We need to release on the error cases below.

yummypeng replied:
updated.

cpuguy83 commented on daemon/commit.go:
```diff
@@ -265,6 +276,10 @@ func (daemon *Daemon) Commit(name string, c *backend.ContainerCommitConfig) (str
 }
 
 func (daemon *Daemon) exportContainerRw(container *container.Container) (io.ReadCloser, error) {
+	rwlayer, err := daemon.layerStore.GetRWLayer(container.ID)
```
We need to release on the error cases below.

yummypeng replied:
updated.

tonistiigi commented on daemon/commit.go:
```diff
@@ -265,7 +276,17 @@ func (daemon *Daemon) Commit(name string, c *backend.ContainerCommitConfig) (str
+			daemon.layerStore.ReleaseRWLayer(rwlayer)
+		}
+	}()
+	err = daemon.Mount(container)
```
Are these mount call even needed anymore?

yummypeng replied:
Why not? I think these mount calls are needed for getting data of container's RWLayer if this container is exited.

cpuguy83 replied:
I don't think it's needed as `TarStream()` should handle this. We just need to take the extra reference.

runcom commented:
ping @yummypeng

yummypeng commented:
Sorry for late update :sweat: 
@aaronlehmann @cpuguy83 @runcom @tonistiigi @dmcgowan PTAL

tonistiigi commented on daemon/commit.go:
```diff
@@ -235,18 +246,18 @@ func (daemon *Daemon) Commit(name string, c *backend.ContainerCommitConfig) (str
 	}
+	defer daemon.stores[container.Platform].layerStore.ReleaseRWLayer(rwlayer)
 
 	archive, err := container.RWLayer.TarStream()
```
`rwlayer.TarStream()`

tonistiigi commented on daemon/export.go:
```diff
@@ -36,7 +47,18 @@ func (daemon *Daemon) ContainerExport(name string, out io.Writer) error {
+		}
+	}()
+
+	err = daemon.Mount(container)
```
I think `rwlayer.Mount()` is better here.

yummypeng replied:
I think it's odd to use `rwlayer.Mount()` directly in `daemon` paclage. But I updated anyway.

yummypeng commented:
ping @aaronlehmann @cpuguy83 @runcom @tonistiigi @dmcgowan PTAL :stuck_out_tongue_winking_eye:

thaJeztah commented:
ping @cpuguy83 @tonistiigi @runcom PTAL

cpuguy83 commented on daemon/export.go:
```diff
@@ -36,7 +47,18 @@ func (daemon *Daemon) ContainerExport(name string, out io.Writer) error {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
```
With this defer, it may be good to use named returns.

yummypeng replied:
updated :heart:

cpuguy83 commented:
Design looks OK.
I'm assuming we have integration tests for this?

tonistiigi commented:
LGTM after defer update

cpuguy83 commented on daemon/commit.go:
```diff
@@ -127,6 +129,15 @@ func (daemon *Daemon) Commit(name string, c *backend.ContainerCommitConfig) (str
 	}
+	if container.IsDead() {
+		err := fmt.Errorf("You cannot commit container %s which is Dead", container.ID)
+		return "", apierrors.NewBadRequestError(err)
```
This will need to be updated to use the conflict error type rather than `apierrors` which is now gone.

cpuguy83 commented on daemon/commit.go:
```diff
@@ -127,6 +129,15 @@ func (daemon *Daemon) Commit(name string, c *backend.ContainerCommitConfig) (str
+
+	if container.IsRemovalInProgress() {
+		err := fmt.Errorf("You cannot commit container %s which is being removed", container.ID)
+		return "", apierrors.NewBadRequestError(err)
```
This will need to be updated to use the conflict error type rather than `apierrors` which is now gone.

cpuguy83 commented:
Also needs a rebase.

yummypeng commented:
updated! @tonistiigi @cpuguy83 PTAL thanks :heart:

cpuguy83 commented on daemon/commit.go:
```diff
@@ -235,18 +246,18 @@ func (daemon *Daemon) Commit(name string, c *backend.ContainerCommitConfig) (str
 	return ioutils.NewReadCloserWrapper(archive, func() error {
-			archive.Close()
-			return container.RWLayer.Unmount()
+			return archive.Close()
```
Do we need this wrapper anymore?

yummypeng replied:
So patient and prudential of you :100: updated!

cpuguy83 commented:
One small thing, but otherwise LGTM.

cpuguy83 reviewed (approved):
LGTM if we can get CI to cooperate.

yummypeng commented:
I don't think CI is relevant:
```bash
13:00:57 Sending build context to Docker daemon   2.56kB
13:00:57 
13:00:57 Step 1/6 : FROM microsoft/windowsservercore
13:00:57  ---> 060f286a19a0
13:00:57 Step 2/6 : RUN mkdir C:\tmp && mkdir C:\busybox
13:00:57  ---> Running in 640f8fe29c3e
13:01:01 hcsshim::UnprepareLayer failed in Win32: The specified network resource or device is no longer available. (0x37) layerId=640f8fe29c3e45da762ec99f8856923ee182294d4f269d31e54e0d4d1528dd37 flavour=1
13:01:01 
13:01:01 
13:01:01 ERROR: Failed 'ERROR: Failed to build busybox image' at 08/20/2017 13:01:01
```

runcom commented:
RS1 keeps failing

vieux commented:
@johnstep can you please take a look at the CI ?

yummypeng commented:
@johnstep I have no idea why CI is always failing when building a image. Could you please help me?

AkihiroSuda commented:
@johnstep PTAL?
```
10:38:13 Step 1/6 : FROM microsoft/windowsservercore
10:38:13  ---> 060f286a19a0
10:38:13 Step 2/6 : RUN mkdir C:\tmp && mkdir C:\busybox
10:38:13  ---> Running in 2fed45390d9a
10:38:18 hcsshim::UnprepareLayer failed in Win32: The specified network resource or device is no longer available. (0x37) layerId=2fed45390d9a68ba8eebc17471d5a698936937e80718c299851eb97b68fba74d flavour=1
10:38:18 
10:38:18 
10:38:18 ERROR: Failed 'ERROR: Failed to build busybox image' at 08/31/2017 10:38:18
```

johnstep commented:
I will look into this today, thanks.

johnstep commented on daemon/commit.go:
```diff
@@ -235,18 +245,15 @@ func (daemon *Daemon) Commit(name string, c *backend.ContainerCommitConfig) (str
 }
 
 func (daemon *Daemon) exportContainerRw(container *container.Container) (io.ReadCloser, error) {
-	if err := daemon.Mount(container); err != nil {
```
Why does this function no longer mount the layer?

yummypeng replied:
We assumed that `TarStream()` should mount the layer if needed. ( see https://github.com/moby/moby/pull/30696#discussion_r117799262 for more information )

But I just noticed that the `Diff()` function for windows requested that the layer should be mounted when calling this function. ( see https://github.com/moby/moby/blob/master/daemon/graphdriver/windows/windows.go#L474 )
So, is this the reason why CI fails all the time?

yummypeng replied:
@johnstep Thanks for finding this. I have modified the codes and now CI on windows machine finally passed but CI for powerpc failed due to a full disk.

yummypeng commented:
CI became :green_heart: eventually. ping @cpuguy83 @tonistiigi PTAL. I have modified codes to address build issue on windows platform.

tonistiigi commented:
Don't want to block this but that comment in windows driver means that the driver doesn't implement the interface correctly and therefore the layerstore interface doesn't work. All other drivers don't have this limitation and automatically mount if needed internally.

johnstep commented:
@tonistiigi Good to know; I will open an issue.

thaJeztah commented:
@cpuguy83 @tonistiigi what's the status here? I see CI is green

@johnstep have you opened an issue? If so, can you link it here?

yummypeng commented:
@thaJeztah I changed my codes to a previous version in order to make CI happy, but @tonistiigi said that the best solution is to implement `Diff()` interface on Windows platform correctly. So I'm waiting for @johnstep to address this issue so that I can update my codes to an unanimous version.

tonistiigi commented:
If needed we can add a `TODO` comment to the mount call(maybe with platform check if it doesn't complicate much) and change the driver in follow-up.

thaJeztah reviewed (approved):
discussed with @tonistiigi and we can merge

LGTM

yummypeng commented:
Thanks for patience of @tonistiigi and @thaJeztah, I have rebased and added a `TODO` comment to the mount call in `exportContainerRw()`. Now CI is :green_heart:

thaJeztah commented:
perfect, thank you!
